### PR TITLE
fix bug 5601 : Grammar_API needs to be refactored.

### PR DESCRIPTION
**Pull request by ghost**
This pull-request merges `Grammar_API` and `API` modules into single module and thus fixing [bug 5601](https://coq.inria.fr/bugs/show_bug.cgi?id=5601).

I think it is OK if this issue is opened and discussed separately.

### CODE REVIEW

**Comment by ejgallego** [comment]
IMHO this should go on top of #808

File: grammar/argextend.mlp [hunk 1]
```diff
@@ -46,17 +46,17 @@ let make_act loc act pil =
   make (List.rev pil)
 
 let make_prod_item = function
-  | ExtTerminal s -> <:expr< Grammar_API.Extend.Atoken (Grammar_API.CLexer.terminal $mlexpr_of_string s$) >>
+  | ExtTerminal s -> <:expr< API.Extend.Atoken (API.CLexer.terminal $mlexpr_of_string s$) >>
```
Comment (ejgallego) [inline grammar/argextend.mlp]: IMO `API` should not be inserted here, similarly for the many occurrences along the patch.

**Review by ejgallego — CHANGES_REQUESTED** [review]
The idea is good, however in my opinion this should go on top of #808 (and thus #788)

File: grammar/argextend.mlp (hunk 1 above)
Comment (matejkosik) [inline grammar/argextend.mlp]: Why?

**Comment by matejkosik** [comment]
Why?

**Comment by Zimmi48** [comment]
> Why? 

The dependency of #808 on #788 looks arbitrary...
However I understand @ejgallego's request to put the present PR on top of #808 because both represent large refactorings and #808 has been delayed for no reason while it solves an issue (the circular dependencies issue, that is also discussed in #851 BTW).

**Comment by matejkosik** [comment]
> The dependency of #808 on #788 looks arbitrary...
> However I understand @ejgallego's request to put the present PR on top of #808 because both represent large refactorings and #808 has been delayed for no reason while it solves an issue (the circular dependencies issue, that is also discussed in #851 BTW).

Maybe it was premature to ask this question. I am actually more interested to understand the reasons for the other objection, i.e. why Emilio thinks that it makes sense to generate unqualified identifiers in macros...

**Comment by ejgallego** [comment]
> The dependency of #808 on #788 looks arbitrary...

It is not.

**Comment by ejgallego** [comment]
> why Emilio thinks that it makes sense to generate unqualified identifiers in macros

Again, because with the current design I'd like to resolve this "at linking time" (this is not technically correct as it really happens at compile time), not a code-writing time. The identifiers are not unqualified.

**Comment by matejkosik** [comment]
> Again, because with the current design I'd like to resolve this "at linking time" (this is not technically correct as it really happens at compile time), not a code-writing time. The identifiers are not unqualified BTW, just API-unqualified.
> 
> I am sure this will evolve in the future, but as of today, it looks like the most sensible choice is to make the open API flag be controlled by the consumers of the library, not by the library itself.

Is it possible to get rid `Constr` or `Term` module in the Coq codebase?

**Comment by ejgallego** [comment]
> Is it possible to get rid of either Constr or Term module in the Coq codebase?
> 
> Is it possible to get rid of Termops module in the Coq codebase?

I would highly favor that.

> I suspect that you think that API is supposed to be some kind of subset of the existing mli files. I don't think that is the case.

Then what it is ? Are you aware of the technical problems wrt to mutual dependencies and ordering of declarations?

**Comment by ejgallego** [comment]
> Why would a plugin writer want to deal with Constr, Term, Termops instead of just, e.g., Term?

Why would a core developer want to deal with that indeed?

**Comment by matejkosik** [comment]
> Why would a core developer want to deal with that indeed?

That is an excellent question. When I asked it, I was told that it helps the core developers to draw a line between the kernel (which is unconditionally part of the trusted computing base) and the rest.

The stuff provided by `Termops` is not necessary for implementing the kernel so core developers prefer not to inflate it and put those things in `Termops` rather than `Term`.

**Comment by Zimmi48** [comment]
> That is an excellent question. When I asked it, I was told that it helps the core developers to draw a line between the kernel (which is unconditionally part of the trusted computing base) and the rest.
>
> The stuff provided by `Termops` is not necessary for implementing the kernel so core developers prefer not to inflate it and put those things in `Termops` rather than `Term`.

Very interesting point. `Termops` is certainly not a very good name choice. `Term_extra` would make it much clearer what the intent is.
@matejkosik What about the distinction between `Term` and `Constr`? Could we get rid of this one at least? (BTW, I would tend naturally to prefer `Term` even if the tradition is `Constr` and this has been reinforced recently with the introduction of `EConstr`. If we decide that we prefer `Term`, then we should get rid completely of the `Constr` terminology including by renaming `EConstr` into `ETerm`. Finally the base type should be `Term.t` or `Constr.t` and not `constr` or `Term.constr`.)

**Comment by matejkosik** [comment]
> Very interesting point. Termops is certainly not a very good name choice. Term_extra would make it much clearer what the intent is.
> @matejkosik What about the distinction between Term and Constr? Could we get rid of this one at least? (BTW, I would tend naturally to prefer Term even if the tradition is Constr and this has been reinforced recently with the introduction of EConstr. If we decide that we prefer Term, then we should get rid completely of the Constr terminology including by renaming EConstr into ETerm. Finally the base type should be Term.t or Constr.t and not constr or Term.constr.)

I agree with you. Saying that, I'll include that as a some of (what I think are) the outstanding problems of the current API.

The reasons why none of them were address at this point is that they are simply too many of them. So the only reasonable way forward is a [gradual refinement](https://github.com/matejkosik/coq/blob/trunk%2BAPI_ROADMAP/API/ROADMAP).

**Comment by Zimmi48** [comment]
> The reasons why none of them were address at this point is that they are simply too many of them. So the only reasonable way forward is a gradual refinement.

I fully agree with this approach but my impression is that on this one, you tried to enforce the "no-aliasing" property too fast, while it could have also been done in gradual refinements.

**Comment by matejkosik** [comment]
> I fully agree with this approach but my impression is that on this one, you tried to enforce the "no-aliasing" property too fast, while it could have also been done in gradual refinements.

Unless I misunderstand myself, I did that in a [separate step](https://github.com/coq/coq/pull/718).

**Comment by matejkosik** [comment]
The reason why `Constr` module was not added to the "API" was that plugins were not using it (why would they?).

**Comment by Zimmi48** [comment]
> Unless I misunderstand myself, I did that in a separate step.

I am aware of that. I'm just saying that this step was a big step which could have been further refined.

**Comment by matejkosik** [comment]
> I am aware of that. I'm just saying that this step was a big step which could have been further refined.

I am curious what you mean. Do you think you can post it in that PR for me to read? It may still be interesting lesson that can be applied for the future activity.

**Comment by Zimmi48** [comment]
I'm simply talking about this discussion about `Term` vs `Constr`. This could have been kept for an independent PR so that proper discussion can happen before a change is made.

File: grammar/argextend.mlp (hunk 1 above)
Comment (gares) [inline grammar/argextend.mlp]: Emilio are you implicitly saying that ARGUMENT EXTEND is used in both main-coq code and plugin code?  If so I understand your remark, if not then I don't see why making generated code more "context dependent" could be a plus.

**Review by gares — APPROVED** [review]

**Comment by matejkosik** [comment]
I just rebased on `trunk` because previous PR's introduced a conflict with this PR
(some of the merged PR updated `API/grammar_API.ml{i,}` files whereas this PR deletes them).

**Comment by ejgallego** [comment]
Superseeded by #905

**Comment by matejkosik** [comment]
Superseeded by #911

**Pull request closed without merging**